### ISSUE TITLE
Slow Maven import and update on large projects

### ISSUE THREAD

**Issue by eliasbalasis** [eclipse-m2e/m2e-core#155]
When importing into Eclipse workspace a Maven repository with a large number on artifacts the process takes too long to complete.

I am having a project with 288 artifacts which takes over 90mins to get imported.

I am suspecting that the larger the project is the more the permutations to be performed by M2E are and so the slower the import process becomes.

This is not productive and the members of my team cannot afford waiting that long each time they need to start a new piece of work.

Can something be done about this? Is there room for tuning and optimization of the M2E processes?

**Comment by mickaelistria** [comment]
Do you have an example of a public maven project that can be used to reproduce this issue?

**Comment by eliasbalasis** [comment]
Unfortunately not, I don't have any public projects big enough to demonstrate the problem.

However, having been observing this issue for a very long time, I have to admit that it feels as if M2E is unnecessarily calculating all possible permutations and the larger the number of Maven artifacts imported in Eclipse workspace the larger the number of permutations that M2E is performing slowing down not only the import but also any M2E rebuild when either changes are made to pom.xml files or when a "Maven / Update" action is performed.

I have to admit though, M2E does a great job building the Maven structure when changes but it is probably calculating too many permutations and perhaps some optimizations could be made to speed it up.

**Comment by laeubi** [comment]
Can you attach a profiler (e.g. JProfiler, Yourkit,VisualVM ...) and share the results where it seem to take most of the time?
This might be related to #123

**Comment by eliasbalasis** [comment]
I am running on Windows 10 and there is no Java executable to monitor, just "eclipse.exe" running and no other processes under it while Maven import or "Maven / Update" is running.

Also, I have been experiencing this problem for years, it has just kept becoming more intensive as complexity of projects increased, meaning number of artifacts and their relationships, transitive dependency chains, dependency baselines, BOMs, use of Maven profiles etc.

**Comment by eliasbalasis** [comment]
Correction,

I had forgotten to assign enough Java heap to Eclipse process.

After adjusting the Eclipse memory settings, the process that used to take over an hour with earlier versions of Eclipse (4.15) now takes just 15 minutes.

Great job and well done.

**Comment by eliasbalasis** [comment]
After having experimented for many years with Eclipse and M2E and with more than enough Java heap asigned to Eclipse process, I am certain that the larger an imported Maven solution is the larger the number of permutations M2E is performing and consequently the longer it takes not only to import the Maven solution the first time but also to process each pom.xml change.

The problem becomes worse when importing a hierarhcy of individual Maven solutions.

Creating a minimal but complex enough project to reproduce the issue is rather impossible as only a project as big as the ones I have been working on could reproduce the issue but unfortunately I cannot share those projects.

I cannot explain the behavior described in my last comment and I consider it erroneous.

I am sure other people have experienced this issue and I hope there are some optimizations or performance improvements that could be made to M2E.

**Comment by eliasbalasis** [comment]
As promised and with reference to #123 

I have collected calls of very large depth on a running Eclipse instance while M2e was processing permutations.
see [snapshot-1637846394151.zip](https://github.com/eclipse-m2e/m2e-core/files/7603226/snapshot-1637846394151.zip) and thread named "Worker-712: Building" similar to the one in #123 but without using "Tycho".

Is this helpful?

**Comment by mickaelistria** [comment]
Can you please clarify what makes you think there are "permutations" involved? I don't get where this comes from.
I have more the impression that m2e code relies too much on `resolveParentProject`, which is an expensive operation, while some other and more efficient strategies can probably be used.

**Comment by eliasbalasis** [comment]
I keep seeing strong repetition of "Building: project: XXX" messages in the "Progress" view often on projects already built over and over, as if possible Maven permutations are being examined continuously.

**Comment by mickaelistria** [comment]
I've looked at the code in more details, and calling "resolveParentProject" could be the root cause. I don't really know (yet?) why project resolution is necessary instead of just project.getParent()...

**Comment by eliasbalasis** [comment]
Thanks @mickaelistria,

this makes a lot of sense and aligns with my observations and thoughts.

I fear we are about to open a can of worms but not unnecessarily as more and more frequently and intensively I am experiencing these delays which gives me the feeling that some expensive operation is continuously and/or unecessarily being called (probably "resolveParentProject"), delaying the whole process.

I am not an M2E expert but let me know if there is anything I can do to help.

**Comment by mickaelistria** [comment]
Submitted https://github.com/eclipse-m2e/m2e-core/issues/155 about that. Let's see if it causes some test failures; if not, I'll probably just merge ASAP to get feedback.

**Comment by mickaelistria** [comment]
I merged a patch that should improve this case. It will be available in snapshots in the next ~15 minutes. Please try upgrading to newer snapshots once they're available and report whether the fix is enough for you.

**Comment by eliasbalasis** [comment]
Thanks @mickaelistria ,

> I am running Eclipse JEE 2021-03 with Spring Boot Tools 4 from "Marketplace" under Windows 10

This implies M2E 1.17.2.20210219-1922

Do you think I should try with latest M2E before trying your fix?

**Comment by eliasbalasis** [comment]
As expected, latest M2E 1.19.0 didn't make any difference.

However, your snapshot is lightning fast.

Where importing into Eclipse a hierarchy of Maven solutions used to take 90mins or more, it now takes <10 mins
Eclipse took <10mins to rebuild all 366 imported Maven artifacts as a result of a change in the topmost ancestor pom.xml

All of my struggles have miraculously disappeared.

I will keep experimenting with this and keep you informed, but excellent progress so far...

**Comment by eliasbalasis** [comment]
I have been experimenting with the snapshot release and I realize that Eclipse has stopped trying to resolve the whole universe all the time.
Instead, Eclipse now has the opportunity to run more tasks in parallel which seems to have a performance impact on my machine but it is well understood and expected while parallelism increase is very much welcome.
I did have a few issues where Eclipse seemed to have frozen a couple of times while trying to save some file changes and I had to forcibly stop Eclipse but it could have been an erroneous situation, I am not 100% convinced.

Overall, both the initial import of the Maven solutions hierarchy (roughly 1000 artifacts) and processing of pom.xml changes have proven to be very fast so far, absolutely incomparable to previous state.

I will keep using the snapshot and let you know if I experience anything unusual...

**Comment by eliasbalasis** [comment]
I keep experimenting quite successfully with the snapshot release.

I did experience Eclipse freezing occasionally for longer that it has normally been acceptable but without recovering this time and I blamed it to my haste making and saving changes while Eclipse was building which in my long experience 
with Eclipse has never been a good approach.
Lesson confirmed once again, if Eclipse is building wait for it.

Overall, the replacement of `resolveParentProject()` with `project.getParent()` seems to be very crucial for the build lifecycle.
However, I didn't understand why project resolution was necessary instead of just `project.getParent()`...
Is it possible that use of `resolveParentProject()` is actually useful, even though my experimentation so far is proving otherwise?

**Comment by mickaelistria** [comment]
> I did experience Eclipse freezing occasionally for longer that it has normally been acceptable

I don't think the change is likely to cause longer freeze; it may be another issue. When this happen, run `jstack` against the Eclipse java process, the stack for the "main" thread would usually show the reason of the freeze, and reading packge names can direct to the culprit project where to report freeze.

**Comment by eliasbalasis** [comment]
Thanks @mickaelistria, I will follow your advice next time.

However, just to be absolutely certain,
is there a possibility that resolveParentProject() is actually useful, even though my experimentation so far is proving otherwise?

**Comment by mickaelistria** [comment]
> is there a possibility that resolveParentProject() is actually useful, even though my experimentation so far is proving otherwise?

I don'y know for sure. I didn't figure out a reasin why readParentProject would be interesting here.

**Comment by dsbanks99** [comment]
@mickaelistria can you confirm, is the no resolveParentProject() enhancement in the latest m2e I would get from the marketplace? If not , how do I get it? I tried pointing my eclipse install to https://download.eclipse.org/technology/m2e/snapshots/1.19.0/latest/ but the latest version seems to be
      M2E - Complete Development Kit (optional)	1.19.0.20211118-0811

**Comment by mickaelistria** [comment]
No, it's not yet released and not accessible on Marketplace. You'll need to install snapshots as described in https://github.com/eclipse-m2e/m2e-core#-installation

**Comment by HannesWell** [comment]
> I tried pointing my eclipse install to https://download.eclipse.org/technology/m2e/snapshots/1.19.0/latest/ but the latest version seems to be M2E - Complete Development Kit (optional) 1.19.0.20211118-0811

Because M2E 1.19.0 was released already the snapshots the repo you used are not updated any more.
To get the latest updates for the release currently under development use (notice `1.19.1` vs. `1.19.0`):
https://download.eclipse.org/technology/m2e/snapshots/1.19.1/latest/
or to always get the latest snapshots use:
https://download.eclipse.org/technology/m2e/snapshots/latest/

**Comment by StrongSteve** [comment]
can anyone point me in the direction of which version of m2e to install to get the speedup during the maven handling?
is it https://download.eclipse.org/technology/m2e/snapshots/1.19.1/latest/?

so basically a version > 1.19.1?
